go/protomodule/protomodule_message.go

```diff
@@ -0,0 +1,370 @@
+		return nil, err
+	}
+	return []byte(jsonData), nil
+}
+
+func (msg *protoMessage) Attr(name string) (starlark.Value, error) {
+	// If a value has already been set on msg, return it
+	if val, ok := msg.fields[name]; ok {
+		return val, nil
+	}
+
+	fieldDesc := getFieldDescriptor(msg.msgDesc, name)
+	if fieldDesc == nil {
+		return starlark.None, fmt.Errorf("AttributeError: `%s' value has no field %q", msg.Type(), name)
+	}
+
+	// Given field name exists but has not been set, return a default value
+	val := fieldDesc.Default()
+
+	starlarkValue, err := valueToStarlark(val, fieldDesc)
+	if err != nil {
+		return starlark.None, err
+	}
+
+	// For non-scalar values, set the value on access even if it is unset so
+	// use without initialization works.
+	//
+	// Example:
+	//   msg = MyProtoMessage()
+	//   msg.repeated_field.append("a")
+	//   # msg.repeated_field should be ["a"]
+	if fieldDesc.IsList() || fieldDesc.IsMap() || fieldDesc.Kind() == protoreflect.MessageKind {
+		msg.SetField(name, starlarkValue)
+	}
+
+	return starlarkValue, nil
+}
+
+func (msg *protoMessage) AttrNames() []string {
+	return fieldNames(msg.msgDesc)
+}
+
+func fieldNames(msgDesc protoreflect.MessageDescriptor) []string {
+	out := []string{}
+	fields := msgDesc.Fields()
+	for i := 0; i < fields.Len(); i++ {
+		fieldName := string(fields.Get(i).Name())
+		out = append(out, fieldName)
+	}
+	sort.Strings(out)
+	return out
+}
+
+func (msg *protoMessage) SetField(name string, val starlark.Value) error {
+	fieldDesc := getFieldDescriptor(msg.msgDesc, name)
+	if fieldDesc == nil {
+		return fmt.Errorf("AttributeError: `%s' value has no field %q", msg.Type(), name)
+	}
+
+	if err := msg.CheckMutable("set field of"); err != nil {
+		return err
+	}
+
+	// Convert starlark.List and starlark.Dict on assignment
+	if fieldDesc.IsList() {
+		if starlarkListVal, ok := val.(*starlark.List); ok {
+			// Convert starlark.List to protoRepeated
+			list, err := newProtoRepeatedFromList(fieldDesc, starlarkListVal)
+			if err != nil {
+				return err
+			}
+
+			val = list
+		}
+	} else if fieldDesc.IsMap() {
+		if starlarkDictVal, ok := val.(*starlark.Dict); ok {
+			// Convert stalark.Map into protoMap
+			mapVal, err := newProtoMapFromDict(fieldDesc.MapKey(), fieldDesc.MapValue(), starlarkDictVal)
+			if err != nil {
+				return err
+			}
+
+			val = mapVal
+		}
+	}
+
+	// Allow using msg_field = None to unset a scalar message field
+	if fieldDesc.Kind() == protoreflect.MessageKind && val == starlark.None && !fieldDesc.IsList() && !fieldDesc.IsMap() {
+		delete(msg.fields, name)
+		return nil
+	}
+
+	// If valueFromStarlark returns an error, the val cannot be assigned to the field
+	_, err := valueFromStarlark(msg.msg.ProtoReflect(), fieldDesc, val)
+	if err != nil {
+		return err
+	}
+
+	// Clear other oneof
+	oneof := fieldDesc.ContainingOneof()
+	if oneof != nil {
+		fields := oneof.Fields()
+		for i := 0; i < fields.Len(); i++ {
+			delete(msg.fields, string(fields.Get(i).Name()))
```

Comment:
I don't fully understand the data structure for one of fields - are the fields in `oneof.Fields()` actually expected to be in `msg.fields`? Or are they in some other data structure nested within `msg.fields`? (Feel free to point me at documentation.)

Reply:
So my understanding of a `oneof` field is that it has a separate `FieldDescriptor` per field, and `oneof.Fields()` returns all the `FieldDescriptors` connected to the same oneof. In our test proto this would be `f_oneof_a` and `f_oneof_b`: https://github.com/stripe/skycfg/blob/trunk/internal/testdata/test_proto/test_proto_v3.proto#L59-L62

In this implementation, if you set any oneof, it clears the other values from the `fields` map (unsetting them) otherwise when `toProtoMessage` iterates over `fields` and calls `msg.Set` on each of them, it will choose the last oneof value set as only one oneof is set and all others are cleared on set

At least in the dynamicpb code this works the same https://github.com/protocolbuffers/protobuf-go/blob/v1.27.1/types/dynamicpb/dynamic.go#L269